encryptionkeyrotation/README.md

```diff
@@ -0,0 +1,95 @@
+
+## Terminology
+
+| Term     | Definition                                                                            |
+| -------- | ------------------------------------------------------------------------------------- |
+| VolumeID | The identifier of the volume generated by the plugin.                                 |
+| CO       | Container Orchestration system that communicates with plugins using CSI service RPCs. |
+| SP       | Storage Provider, the vendor of a CSI plugin implementation.                          |
+| RPC      | [Remote Procedure Call](https://en.wikipedia.org/wiki/Remote_procedure_call).         |
+
+## Objective
+
+Define an extension to the CSI specification that will enable storage vendors
+(SP) to develop controllers/plugins that can rotate the Encryption Keys (KEK)
+of existing volumes that are encrypted.
+
+### Goals in MVP
+
+The new extension will define a procedure that
+
+- can be called for existing encrypted volumes
+- functions for Block mode volumes
+- interacts with the Node-Plugin
+- makes it possible for the SP to rotate encrpytion keys on volumes
+
+### Non-Goals in MVP
+
+- Rotating encrpytion keys for inactive volumes
+
+## Solution Overview
+
+This specification defines an interface along with the minimum operational and
+packaging recommendations for a storage provider (SP) to implement a key rotate operation for volumes. The interface declares the RPCs that a plugin
+MUST expose.
+
+## RPC Interface
+
+- **Node Service**: The Node plugin MAY implement this RPC.
+
+```protobuf
+syntax = "proto3";
+package encryptionkeyrotation;
+
+import "github.com/container-storage-interface/spec/lib/go/csi/csi.proto";
+import "google/protobuf/descriptor.proto";
+
+option go_package = "github.com/csi-addons/spec/lib/go/encryptionkeyrotation";
+
+// EncryptionKeyRotationController holds the RPC method for running
+// key rotation operation on the volume.
+service EncryptionKeyRotationController {
+    // EncryptionKeyRotate is a procedure that is called on the CSI Nodeplugin
+    rpc EncryptionKeyRotate(EncryptionKeyRotateRequest)
+    returns (EncryptionKeyRotateResponse){}
+}
+```
+
+### EncryptionKeyRotateRequest
+
+```protobuf
+// EncryptionKeyRotateRequest contains the information needed to identify
+// the volume by the SP and access any backend services so that the key can be
+// rotated.
+message EncryptionKeyRotateRequest {
+    // The ID of the volume for which the key is to be rotated. This field is REQUIRED.
+    string volume_id = 1;
+
+    // Plugin specific parameters passed in as opaque key-value pairs.
+    map<string, string> parameters = 2;
+
+    // Secrets required by the plugin to complete the request.
+    map<string, string> secrets = 3 [(csi.v1.csi_secret) = true];
+}
+```
+
+### EncryptionKeyRotateResponse
+
+```protobuf
+// EncryptionKeyRotateResponse holds the information about the result of the
+// EncryptionKeyRotateRequest call.
+message EncryptionKeyRotateResponse {
+
```

Comment:
what *information about the result* will this contain?

Reply:
Well nothing actually, if it doesn't return error means it was successful. We can see details about it using the status field on the CR.

Reply:
If you want to put details in a status field in a (Kubernetes) CR, those details need to be passed from the CSI-driver to the kubernetes-csi-addons controller. The contents of this gRPC message would be the place where it should be recorded.

Reply:
The status will be one of `started`, `success` and `failed`. All of which can be determined with the return codes.